### PR TITLE
IEP 293 allow to enter project name in the templates wizard (IEP-295)

### PR DESCRIPTION
**Pull request by sigmaaa**
Added project name text field with the label. When the option to select a template is available, the project name is based on the selected template. Validation of project name based on validation from main page. Few cases of not passed validation:

- **Case 1** - spaces in the name :

IEP-293

- **Case 2** - project name already exists:

IEP-293case 2

When validation is not passed successfully Finish button is unavailable. 
Also, when the project name is changed on the templates page it will be also changed on the main page.

### CODE REVIEW

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/AbstractTemplatesSelectionPage.java [hunk 1]
```diff
@@ -180,16 +180,23 @@ public void selectionChanged(SelectionChangedEvent event)
...
 			e.printStackTrace();//TODO log the exception
 		}
+		setProjectName(projectName);
 		setDescriptionText(description);
 		getContainer().updateButtons();
 	}
+	
+	public void setProjectName(String projectName) 
+	{
+	}
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/AbstractTemplatesSelectionPage.java:199]: Can we remove the empty method?

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/TemplateListSelectionPage.java [hunk 2]
```diff
@@ -8,23 +8,39 @@
 
 import java.io.IOException;
 
+import org.eclipse.core.internal.resources.LocalMetaArea;
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/TemplateListSelectionPage.java:11]: Remove unused imports by cleaning up the code

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/TemplateListSelectionPage.java [hunk 3]
```diff
@@ -8,23 +8,39 @@
...
+import org.eclipse.swt.events.ModifyListener;
 import org.eclipse.swt.layout.GridData;
+import org.eclipse.swt.layout.GridLayout;
+import org.eclipse.swt.layout.RowLayout;
 import org.eclipse.swt.widgets.Button;
 import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Group;
+import org.eclipse.swt.widgets.Label;
+import org.eclipse.swt.widgets.Text;
+import org.eclipse.ui.internal.ide.IDEWorkbenchMessages;
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/TemplateListSelectionPage.java:28]: Can you remove these internal workbench messages usage

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/TemplateListSelectionPage.java [hunk 4]
```diff
@@ -8,23 +8,39 @@
...
+import org.eclipse.swt.events.ModifyEvent;
+import org.eclipse.swt.events.ModifyListener;
 import org.eclipse.swt.layout.GridData;
+import org.eclipse.swt.layout.GridLayout;
+import org.eclipse.swt.layout.RowLayout;
 import org.eclipse.swt.widgets.Button;
 import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Group;
+import org.eclipse.swt.widgets.Label;
+import org.eclipse.swt.widgets.Text;
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/TemplateListSelectionPage.java:27]: There a couple of unused imports, please clean up

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/TemplateListSelectionPage.java [hunk 5]
```diff
@@ -80,17 +131,30 @@ private void selectInitialTemplate() throws IOException
...
 	{
-		return true; // will always to finish the page without template selection also
+		return isPageCreated;
 	}
 
+	@Override
+	public void selectionChanged(SelectionChangedEvent event) 
+	{
+		super.selectionChanged(event);
+	}
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/TemplateListSelectionPage.java:147]: As there is no change in the default implementation, I think we can remove it

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/wizard/WizardNewProjectCreationPage.java [hunk 6]
```diff
@@ -20,6 +20,7 @@
 
 import java.net.URI;
 
+import org.eclipse.cdt.internal.core.language.settings.providers.LanguageSettingsExtensionManager;
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/wizard/WizardNewProjectCreationPage.java:23]: I think unused import, we can remove it

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/wizard/WizardNewProjectCreationPage.java [hunk 7]
```diff
@@ -53,14 +56,17 @@
...
 	private String initialProjectFieldValue;
 
+	private TemplateListSelectionPage templatesPage;
 	// widgets
 	Text projectNameField;
 
 	private Listener nameModifyListener = e -> {
+		if (templatesPage.isCreated()) {
+			templatesPage.setProjectName(projectNameField.getText());
+		}
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/wizard/WizardNewProjectCreationPage.java:66]: The better way to get the project name in the TemplateListSelectionPage is to make use of the eclipse wizard **getPreviousPage()** instance and get the name. Can you explore more on this?

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/TemplateListSelectionPage.java [hunk 8]
```diff
@@ -143,5 +207,9 @@ public void setVisible(boolean visible)
 		}
 		super.setVisible(visible);
 	}
-
+	
+	public void setMainPage(WizardNewProjectCreationPage mainPage) 
+	{
+		this.mainPage = mainPage;
+	}
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/TemplateListSelectionPage.java:214]: We can avoid this method if we use getPreviousPage() and access the required methods/calls

**Review by kolipakakondal — CHANGES_REQUESTED** [review]
Hi @sigmaaa 

Could you please improve on the review comments. It seems to be there are issues with the name modify listener as well. Please take a look again.

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/TemplateListSelectionPage.java [hunk 9]
```diff
@@ -134,6 +188,8 @@ public void setInitialTemplateId(ITemplateNode templateNode)
 	@Override
 	public void setVisible(boolean visible)
 	{
+
+		projectNameField.setText(prevPage.getProjectName());
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/TemplateListSelectionPage.java:192]: Can we set this in the **createProjectNameGroup**() itself as this is a one-time activity, and importantly this call is inappropriate for the **setVisible**() method

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/TemplateListSelectionPage.java [hunk 10]
```diff
@@ -80,15 +129,20 @@ private void selectInitialTemplate() throws IOException
...
-
 		String description = new TemplatesManager().getDescription(getInitialTemplateId());
 		setDescriptionText(description);
 	}
 
 	@Override
-	public boolean isPageComplete()
+	public void selectionChanged(SelectionChangedEvent event)
 	{
-		return true; // will always to finish the page without template selection also
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/TemplateListSelectionPage.java:91]: If this is removed, are we allowing to create a default template project even without a selection?

**Review by kolipakakondal — CHANGES_REQUESTED** [review]
Hi @sigmaaa Could you address the pointed comments. Otherwise, PR looks good, we can go ahead with the merge once that's addressed.

**Comment by kolipakakondal** [comment]
Hi @sigmaaa Once the specific comment/issue is resolved click on the "**Resolve Conversation**" button

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/TemplateListSelectionPage.java (hunk 10 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/templates/TemplateListSelectionPage.java:91]: I forgot to add it after changes, thank you for the point. Returned it back.

**Review by kolipakakondal — APPROVED** [review]
LGTM

**Pull request merged**